File: python/linguistic_variation_toolbox/_private/compute_category_statistics.py
from __future__ import annotations

import math
import numpy as np
import networkx as nx
import pandas as pd

from .is_category_reference_in import is_category_reference_in


def _cmdscale(D: np.ndarray) -> np.ndarray:
    """Classical metric MDS matching MATLAB's cmdscale()."""
    n = D.shape[0]
    J = np.eye(n) - np.ones((n, n)) / n
    B = -0.5 * J @ (D ** 2) @ J
    eigenvalues, eigenvectors = np.linalg.eigh(B)
    idx = np.argsort(eigenvalues)[::-1]
    eigenvalues = eigenvalues[idx]
    eigenvectors = eigenvectors[:, idx]
    pos = eigenvalues > 1e-10
    return eigenvectors[:, pos] * np.sqrt(eigenvalues[pos])


def _compute_distance_from_baricentre(subgraph: nx.Graph) -> list[float]:
    num_nodes = subgraph.number_of_nodes()
    if num_nodes <= 1:
        return [0.0] * num_nodes
    D = nx.to_numpy_array(subgraph, weight="Weight")
    coords = _cmdscale(D)
    baricentre = coords.mean(axis=0)
    centered = coords - baricentre
    return list(np.linalg.norm(centered, axis=1))


def _compute_weight_range(subgraph: nx.Graph) -> list[float]:
    nodes = list(subgraph.nodes)
    num_nodes = len(nodes)
    if num_nodes <= 1:
        return [float("nan")] * num_nodes
    result = []
    for node in nodes:
        weights = [d["Weight"] for _, _, d in subgraph.edges(node, data=True)]
        if not weights:
            result.append(float("nan"))
        else:
            result.append(max(weights) - min(weights))
    return result


def _compute_variant_stats(subgraph: nx.Graph) -> pd.DataFrame:
    nodes = list(subgraph.nodes)
    weights_arr = [d["Weight"] for _, _, d in subgraph.edges(data=True)]
    max_w = max(weights_arr) if weights_arr else 1.0
    inverse_weights = {(u, v): max_w - d["Weight"] + 1 for u, v, d in subgraph.edges(data=True)}

    nx.set_edge_attributes(subgraph, inverse_weights, "cost")

    num_variants = len(nodes)
    weighted_degree = []
    for node in nodes:
        wd = sum(d["Weight"] for _, _, d in subgraph.edges(node, data=True))
        weighted_degree.append(wd)

    mean_distance = [wd / (num_variants - 1) if num_variants > 1 else float("nan")
                     for wd in weighted_degree]

    range_distance = _compute_weight_range(subgraph)
    distance_from_baricentre = _compute_distance_from_baricentre(subgraph)

    closeness_raw = nx.closeness_centrality(subgraph, distance="cost")
    closeness = [closeness_raw[n] for n in nodes]

    attrs = [subgraph.nodes[n]["Attributes"] for n in nodes]
    is_ref = [subgraph.nodes[n]["IsCategoryReference"] for n in nodes]

    df = pd.DataFrame({
        "Name": nodes,
        "Attributes": attrs,
        "IsCategoryReference": is_ref,
        "WeightedDegree": weighted_degree,
        "MeanDistance": mean_distance,
        "RangeDistance": range_distance,
        "Closeness": closeness,
        "DistanceFromBaricentre": distance_from_baricentre,
    })
    return df.sort_values("WeightedDegree", ascending=True).reset_index(drop=True)


def _compute_general_stats(subgraph: nx.Graph, distance_from_baricentre: list[float]) -> dict:
    distances = [d["Weight"] for _, _, d in subgraph.edges(data=True)]
    stats: dict = {}
    if not distances:
        stats["Diameter"] = 0
        stats["MeanDistance"] = float("nan")
        stats["RangeDistance"] = float("nan")
        stats["MeanDistanceFromBaricentre"] = 0.0
    else:
        stats["Diameter"] = max(distances)
        stats["MeanDistance"] = float(np.mean(distances))
        stats["RangeDistance"] = max(distances) - min(distances)
        stats["MeanDistanceFromBaricentre"] = float(np.mean(distance_from_baricentre))
    stats["NumVariants"] = subgraph.number_of_nodes()
    return stats


def compute_category_statistics(graph: nx.Graph, category: str | None) -> dict:
    """Compute statistics for all variants in a category (or the whole graph if category is None)."""
    if category is not None:
        category_nodes = [
            n for n in graph.nodes
            if any(a.category == category for a in graph.nodes[n]["Attributes"])
        ]
        if not category_nodes:
            return {}
        subgraph = graph.subgraph(category_nodes).copy()
        ref_flags = is_category_reference_in(subgraph, category)
        for node, flag in zip(list(subgraph.nodes), ref_flags):
            subgraph.nodes[node]["IsCategoryReference"] = flag
    else:
        subgraph = graph

    if subgraph.number_of_nodes() == 0:
        return {}

    variant_data = _compute_variant_stats(subgraph)
    stats = _compute_general_stats(subgraph, list(variant_data["DistanceFromBaricentre"]))
    stats["VariantData"] = variant_data
    return stats

File: matlab/source/private/computeCategoryStatistics.m
function [stats, categorySubGraph] = computeCategoryStatistics(inputGraph, category)
%COMPUTECATEGORYSTATISTICS Compute some statistics on the data given the
%category.

% Copyright 2023 Acadèmia de su Sardu APS

if ~isempty(category)
    isCategory = cellfun(@(a) iIsCurrentCategory(a, category), ...
        inputGraph.Nodes.Attributes);
    nodeIndices = find(isCategory);
    categorySubGraph = subgraph(inputGraph, nodeIndices);
    categorySubGraph.Nodes.IsCategoryReference = ...
        isCategoryReferenceIn(categorySubGraph, category);
else
    categorySubGraph = inputGraph;
end

if categorySubGraph.numnodes()==0
    stats = struct();
    return
end

variantData = iComputeVariantStats(categorySubGraph);

stats = iComputeGeneralStats(categorySubGraph, variantData.DistanceFromBaricentre);
stats.VariantData = variantData;
end


function tf = iIsCurrentCategory(attributes, category)
tf = false;
if any([attributes.Category]==category)
    tf = true;
end
end


function variantStats = iComputeVariantStats(categorySubGraph)
variantStats = categorySubGraph.Nodes;
weights = categorySubGraph.Edges.Weight;
numVariants = categorySubGraph.numnodes();
inverseWeights = max(weights) - weights + 1;

variantStats.WeightedDegree = centrality(categorySubGraph, 'degree', Importance=weights);
variantStats.MeanDistance = variantStats.WeightedDegree./(numVariants-1);
variantStats.RangeDistance = iComputeWeightRange(categorySubGraph);
variantStats.Closeness = centrality(categorySubGraph, 'closeness', Cost=inverseWeights);
variantStats.DistanceFromBaricentre = iComputeDistanceFromBaricentre(categorySubGraph);
variantStats = sortrows(variantStats, "WeightedDegree", "ascend");
end


function generalStats = iComputeGeneralStats(categorySubGraph, distanceFromBaricentre)
generalStats = struct();
distances = categorySubGraph.Edges.Weight;
if isempty(distances)
    generalStats.Diameter = 0;
    generalStats.MeanDistance = nan;
    generalStats.RangeDistance = nan;
    generalStats.MeanDistanceFromBaricentre = 0;
else
    generalStats.Diameter = max(distances, [], "all");
    generalStats.MeanDistance = mean(distances, "all");
    generalStats.RangeDistance = range(distances, "all");
    generalStats.MeanDistanceFromBaricentre = mean(distanceFromBaricentre, "all");
end
generalStats.NumVariants = numnodes(categorySubGraph);
end


function outRange = iComputeWeightRange(aGraph)
numNodes = aGraph.numnodes();
if numNodes<=1
    outRange = nan(numNodes, 1);
    return
end

weights = aGraph.Edges.Weight;
outRange = zeros([numNodes, 1]);
[startNodes, endNodes] = aGraph.findedge();
for k = 1:numNodes
    currEdges = startNodes==k | endNodes==k;
    outRange(k) = range(weights(currEdges));
end
end

function outDistance = iComputeDistanceFromBaricentre(aGraph)
distanceMatrix = aGraph.adjacency('weighted');
numNodes = size(distanceMatrix, 1);
if numNodes<=1
    outDistance = 0;
    return
end

coordinates = cmdscale(full(distanceMatrix));
baricentre = mean(coordinates);
centeredCoordinates = coordinates-baricentre;
outDistance = vecnorm(centeredCoordinates, 2, 2);
end